course/viewbase.py

```diff
@@ -57,31 +57,43 @@ def get_resource_objects(self):
             self.instance = instance
             self.course = self.instance.course
             self.content = CachedContent(self.instance)
+            self.user_in_course = None
+            if user.is_authenticated:
+                self.user_in_course = self.instance.get_enrollment_for(user)
             self.is_student = self.instance.is_student(user)
             self.is_assistant = self.instance.is_assistant(user)
             self.is_teacher = self.course.is_teacher(user)
             self.is_course_staff = self.is_teacher or self.is_assistant
             self.get_taggings = lambda: CachedStudent(instance, user.id).data['tag_slugs']
 
             self.note(
-                "course", "instance", "content", "is_student", "is_assistant",
+                "course", "instance", "content", "user_in_course", "is_student", "is_assistant",
                 "is_teacher", "is_course_staff", "get_taggings",
             )
 
-            # Apply course instance language.
+            # Try to find a language that is defined for this course instance
+            # and apply it
             if self.instance.language:
-                lang = self.instance.language
-                if lang.startswith("|"):
-                    active = get_language()
-                    if "-" in active:
-                        active = active.split("-")[0]
-                    if "|" + active + "|" in lang:
-                        translation.activate(active)
-                    else:
-                        fallback = lang[1:lang.find("|", 1)]
-                        translation.activate(fallback)
+                instance_languages = self.instance.language
+
+                languages = []
+                if self.user_in_course and self.user_in_course.language:
+                    languages.append(self.user_in_course.language)
+                if user.is_authenticated and user.userprofile.language:
+                    languages.append(user.userprofile.language)
+                languages.append(get_language())
+
+                for lang in languages:
+                    lang = lang.split("-")[0]
```

Comment:
If `languages = ['en-fi', 'fi']` and ` instance_languages = |en|sv|`, then this code will activate language `translation.activate('en')`, which would mean that formats strored in `formats/en-fi/formats.py` would not be used.

Thus, I think you should do:

```python3
instance_languages = self.instance.language.strip('|').split('|')
instance_def_language = instance_languages[0]
instance_languages = set(instance_languages) # find (i.e. `in`) is fast with sets

for lang in languages:
  if lang.split('-', 1)[0] in instance_languages:
    language = lang
    break
```

However, this requires to fix all instances of `get_language()` in rest of the code for `en-fi` etc. to work correctly, which of course would be a problem for another issue/PR.

Reply:
This assumes that courses can use only the language part of the locale string (i.e. they can't use `en-us`).

Reply:
Good points regarding the `formats/en-fi/formats.py` implementation.

> This assumes that courses can use only the language part of the locale string (i.e. they can't use en-us).

Should we try to handle this somehow or is it okay to assume that the courses can now use only the language part?

Reply:
> -- is it okay to assume that the courses can now use only the language part?

Yes, I think we can do that assumption.

p.s. sorry, missed the comment when handling notifications.